substrate-archive/src/actors/workers/state_tracing.rs

```diff
@@ -0,0 +1,566 @@
+/// should be all the traces produces from the execution of a single block.
+pub struct TracingActor<B: BlockT> {
+	targets: Vec<(String, Level)>,
+	database: Address<ActorPool<super::DatabaseActor<B>>>,
+}
+
+impl<B: BlockT> TracingActor<B> {
+	pub fn new(targets: String, database: Address<ActorPool<super::DatabaseActor<B>>>) -> Self {
+		let targets = targets.split(',').map(|s| parse_target(s)).collect();
+		TracingActor { targets, database }
+	}
+
+	/// Returns true if a span is part of an enabled Target. Checks WASM in addition to the spans target.
+	fn is_enabled(&self, span: &SpanMessage) -> bool {
+		let wasm_target = span.values.0.get(WASM_TARGET_KEY).map(|s| s.to_string());
+
+		self.targets.iter().filter(|t| t.0.as_str() != "wasm_tracing").any(|t| {
+			let check_target = |target: &str, lvl: Level| -> bool { target.starts_with(&t.0.as_str()) && lvl <= t.1 };
+
+			check_target(&span.target, span.level)
+				|| wasm_target.as_ref().map(|wasm_t| check_target(wasm_t, span.level)).unwrap_or(false)
+		})
+	}
+
+	/// Formats spans based upon data types that would be more useful for querying in the context
+	/// of a relational database.
+	fn format_spans(&self, spans: Vec<SpanMessage>) -> Result<Vec<SpanMessage>> {
+		let format = |mut span: SpanMessage| -> Result<SpanMessage> {
+			if span.name == WASM_TRACE_IDENTIFIER {
+				if let Some(name) = span.values.0.remove(WASM_NAME_KEY) {
+					span.name = name.to_string();
+				}
+				if let Some(target) = span.values.0.remove(WASM_TARGET_KEY) {
+					span.target = target.to_string();
+				}
+				span.file = span.values.0.remove("file").map(Into::into);
+				span.line = match span.values.0.remove("line") {
+					Some(DataType::U64(t)) => Ok(Some(t.try_into()?)),
+					None => Ok(None),
+					_ => Err(TracingError::TypeError),
+				}?;
+			}
+			Ok(span)
+		};
+		// if we have a different name/target from WASM replace it and remove key from TraceData
+		spans.into_iter().filter(|s| s.name != BLOCK_EXEC_SPAN && self.is_enabled(&s)).map(format).collect()
+	}
+
+	/// Handles a single event message.
+	async fn handle_event(&self, event: EventMessage) -> Result<()> {
+		self.database.send(event.into()).await?;
+		Ok(())
+	}
+
+	/// Handles CollatedSpans. Collated spans are all spans that result from the execution of a block.
+	async fn handle_spans(&mut self, spans: CollatedSpans) -> Result<()> {
+		let block_num = spans.block_num().ok_or(TracingError::NoBlockNumber)?;
+		let hash = spans.hash().ok_or(TracingError::NoHash)?;
+		let spans = self.format_spans(spans.into_inner())?;
+		self.database.send(Traces::new(block_num, hash, spans).into()).await?;
+		Ok(())
+	}
+}
+
+#[async_trait::async_trait]
+impl<B: BlockT> Actor for TracingActor<B> {
+	async fn started(&mut self, ctx: &mut Context<Self>) {
+		let addr = ctx.address().expect("Actor just started");
+		let handler = ArchiveTraceHandler::new(addr, self.targets.clone());
+		if tracing::subscriber::set_global_default(handler).is_err() {
```

Comment:
It's possible to limit the scope of a subscriber to a closure, which might make it easier to group traces by block, rather than setting a global default:

https://docs.rs/tracing/0.1.22/tracing/dispatcher/fn.with_default.html